GoogleForCreators/web-stories-wp
# Custom Fonts: Allow inserting via URL

## Context
Adds the user-facing side of adding custom fonts.


## Summary
Adds Custom Fonts section to the Dashboard (behind a feature flag)
Displays the added custom fonts in the font picker.


## Relevant Technical Choices



## To-do

- [x] Use actual REST API calls instead of static data
- [ ] Finalize UX and use better messaging (Blocked -- UX is not finalized + messaging is also WIP)
- [x] Add tracking for adding a custom font
- [x] Karma updates



## User-facing changes
<img width="292" alt="Screenshot 2021-12-15 at 11 40 28" src="https://user-images.githubusercontent.com/3294597/146162106-8a74762d-297e-4ad0-a245-fe42f3f56788.png">
<img width="897" alt="Screenshot 2021-12-15 at 11 49 47" src="https://user-images.githubusercontent.com/3294597/146163704-b78c90a8-5c21-40dc-82d2-91b626c36167.png">




## Testing Instructions







This PR can be tested by following these steps:

Note: adding fonts can fail due to CORS issues, if the user wants to use that font, they need to configure the settings from where the font is hosted. This is ok for MVP.

1. Enable Custom Fonts feature
2. Go to Dashboard, try adding a font with some random/incorrect URLs, verify it doesn't work, and error messages are displayed.
3. Try with a working font URL: .otf / .ttf / .woff, e.g. https://overpass-30e2.kxcdn.com/overpass-regular.ttf
4. Verify the font gets added to the list.
5. Verify it's possible to delete the font.
6. Make sure that you have some fonts added and open the editor.
7. Add a text and open the Font Picker. Verify the custom fonts are displayed in the font picker as the first group.
8. Verify you can use the fonts and it also displays in the front-end.


## Reviews

### Does this PR have a security-related impact?
Adds a new feature that might beed configuring CORS settings by the user.


### Does this PR change what data or activity we track or use?
Adds tracking adding a custom font by event `'add_custom_font'` and `url` as additional data.


### Does this PR have a legal-related impact?
No


## Checklist




---



Fixes #9880
